include/deal.II/hp/fe_collection.h

```diff
@@ -371,6 +372,32 @@ namespace hp
     find_dominating_fe_in_subset(const std::set<unsigned int> &fes,
                                  const unsigned int            codim = 0) const;
 
+    /**
+     * Set functions determining the hierarchy of finite elements, i.e. a
+     * function @p next that returns the index of the finite element following
+     * the given one, and a function @p prev returning the preceding one.
+     *
+     * @note Both passed and returned indices have to be valid within the index
+     * range of this collection, i.e. within [0, size()).
+     */
+    void
+    set_hierarchy(const std::function<unsigned int(const unsigned int)> &next,
+                  const std::function<unsigned int(const unsigned int)> &prev);
```

Comment:
I would pass the whole `FECollection` as first argument to these functions. Then they can be global functions, for example, that don't need to know about other pieces of the program.

Reply:
There probably also ought to be a `clear_hierarchy()` function. You can then call that from the constructor, because it does what you currently do in the constructor.